EIPS/eip-2315.md

```diff
@@ -13,55 +13,67 @@ requires: 3540, 3670, 4200
 
-We believe this to be one of the simplest possible proposals that meets these requirements
+Valid code will support streaming, one-pass, and other near-linear compilers.
+
+This is among the simplest possible proposals that meets these requirements.
 
 ## Motivation
 
-### We want a complete control-flow facility.
+### We need a complete control-flow facility.
 
 Jumps, conditional jumps and subroutines were proposed by Alan Turing in 1945 as a means of organizing the logic of the code and the design of the memory crystals for his Automatic Computing Engine:
 > "We wish to be able to arrange that sequences of orders can divide at various points, continuing in different ways according to the outcome of the calculations to date...  We also wish to be able to arrange for the splitting up of operations into subsidiary operations... To start on a subsidiary operation we need only make a note of where we left off the major operation and then apply the first instruction of the subsidiary.  When the subsidiary is over we look up the note and continue with the major operation."
 > 
-> — Alan Turing — *cf. B.E. Carpenter, R.W. Doran, "The other Turing machine." The Computer Journal, Volume 20, Issue 3, January 1977.*
+> — Alan Turing — in [B.E. Carpenter, R.W. Doran, "The other Turing machine." The Computer Journal, Volume 20, Issue 3, DOI: 10.1093/comjnl/20.3.269, January 1977.](https://academic.oup.com/comjnl/article/20/3/269/751954)
+
+In more contemporary terms, we have sequences of instructions, jumps and conditional jumps that divide sequences into blocks, subroutine calls, and a stack of addresses to return to.  The details vary, but similar facilities have proven their value across a long line of important machines over the last 75 years, from the ACE to the Cray to the ARM.
 
-In more contemporary terms, we have sequences of instructions, jumps and conditional jumps that divide sequences into blocks, subroutine calls, and a stack of addresses to return to.  The details vary, but similar facilities have proven their value across a long line of machines going back over 75 years.
+Unlike these machines, the Ethereum Virtual Machine _does not_ provide subroutine operations.  Instead, they must be synthesized using the dynamic `JUMP` instruction, which takes its destination on the stack. Further, the EVM provides _only_ dynamic jumps, impeding the static analysis we need.
 
-Unlike these machines, the Ethereum Virtual Machine _does not_ provide subroutine operations.  Instead, they must be synthesized using the dynamic `JUMP` instruction, which takes its destination on the stack. Further, the EVM provides _only_ dynamic jumps, preventing the static analysis we need.
+### We need safe control-flow.
```

Comment:
```suggestion
###  The control-flow needs to be safer
```

Optional, just to see if this way puts more emphasize on the "safer"

Reply:
I think I sort of weasel on "safe" versus "safer" throughout.  Perfect safety is impossible.